zcutil/make-release.py

```diff
@@ -0,0 +1,529 @@
+        '--repo',
+        dest='REPO',
+        type=str,
+        help='Path to repository root.',
+    )
+    p.add_argument(
+        'RELEASE_VERSION',
+        type=Version.parse_arg,
+        help='The release version: vX.Y.Z-N',
+    )
+    p.add_argument(
+        'RELEASE_PREV',
+        type=Version.parse_arg,
+        help='The previously released version.',
+    )
+    p.add_argument(
+        'RELEASE_HEIGHT',
+        type=int,
+        help='A block height approximately occuring on release day.',
+    )
+    return p.parse_args(args)
+
+
+# Top-level flow:
+def main_logged(release, releaseprev, releaseheight):
+    verify_releaseprev_tag(releaseprev)
+    initialize_git(release)
+    patch_version_in_files(release, releaseprev)
+    patch_release_height(releaseheight)
+    commit('Versioning changes.')
+
+    build()
+    gen_manpages()
+    commit('Updated manpages.')
+
+    gen_release_notes(release)
+    update_debian_changelog(release)
+    commit('Updated release notes and changelog.')
+
+
+def phase(message):
+    def deco(f):
+        @wraps(f)
+        def g(*a, **kw):
+            logging.info('%s', message)
+            return f(*a, **kw)
+        return g
+    return deco
+
+
+@phase('Checking RELEASE_PREV tag.')
+def verify_releaseprev_tag(releaseprev):
+    candidates = []
+    for tag in sh_out('git', 'tag', '--list').splitlines():
+        if tag.startswith('v1'):  # Ignore v0.* bitcoin tags and other stuff.
+            candidates.append(Version.parse_arg(tag))
+
+    candidates.sort()
+    try:
+        latest = candidates[-1]
+    except IndexError:
+        raise SystemExit('No previous releases found by `git tag --list`.')
+
+    if releaseprev != latest:
+        raise SystemExit(
+            'The latest candidate in `git tag --list` is {} not {}'
+            .format(
+                latest.vtext,
+                releaseprev.vtext,
+            ),
+        )
+
+
+@phase('Initializing git.')
+def initialize_git(release):
+    junk = sh_out('git', 'status', '--porcelain')
+    if junk.strip():
+        raise SystemExit('There are uncommitted changes:\n' + junk)
+
+    branch = sh_out('git', 'rev-parse', '--abbrev-ref', 'HEAD').strip()
+    if branch != 'master':
+        raise SystemExit(
+            "Expected branch 'master', found branch {!r}".format(
+                branch,
+            ),
+        )
+
+    logging.info('Pulling to latest master.')
+    sh_log('git', 'pull', '--ff-only')
+
+    branch = 'release-' + release.vtext
+    logging.info('Creating release branch: %r', branch)
+    sh_log('git', 'checkout', '-b', branch)
+    return branch
+
+
+@phase('Patching versioning in files.')
+def patch_version_in_files(release, releaseprev):
+    patch_README(release, releaseprev)
+    patch_clientversion_h(release)
+    patch_configure_ac(release)
+    patch_gitian_linux_yml(release, releaseprev)
+
+
+@phase('Patching release height for auto-senescence.')
+def patch_release_height(releaseheight):
```

Comment:
The 'auto-sensence' feature design, as I imagined it, is 'sometime after a release is deprecated it will automatically halt'. The 'sometime after' was deliberate wording to let us rely on block heights and conservative estimates even though the deprecation policy is in calendar time. In terms of protocol planning we can rely on block heights. Humans have a calendar deadline even though it's a conservative fiction.